src/tiledbsoma_ml/query_ids.py

```diff
@@ -0,0 +1,198 @@
+from tiledbsoma_ml._utils import batched, splits
+from tiledbsoma_ml.common import NDArrayJoinId
+
+logger = logging.getLogger("tiledbsoma_ml.query_ids")
+
+
+Chunks = List[NDArrayJoinId]
+r"""Return-type of |QueryIDs.shuffle_chunks|, |List| of |ndarray|\ s."""
+SamplingMethod = Literal["deterministic", "multinomial", "stochastic_rounding"]
+r"""Enum arg to |QueryIDs.split|:
+
+- ``"deterministic"``: number of each class returned will always be :math:`frac \times N`, rounded to nearest int, e.g.
+  ``n=12, fracs=[.7,.3]`` will always produce 8 and 4 elements, resp.
+- ``"multinomial"``: each element is assigned to a class independently; no guarantees are made about resulting class
+  sizes.
+- ``"stochastic_rounding"``: guarantee each class gets assigned at least :math:`\lfloor frac \times N \rfloor` elements.
+  The remainder are then distributed so that class-size expected-values match the provided ``fracs``.
+"""
+SamplingMethods = ["deterministic", "multinomial", "stochastic_rounding"]
+"""Possible values of |SamplingMethod|."""
+
+
+@define(frozen=True)
+class Partition:
+    rank: int
+    """GPU-process rank."""
+    world_size: int
+    """Number of GPU processes."""
+    worker_id: int
+    """|DataLoader|-worker index."""
+    n_workers: int
+    """Number of |DataLoader|-workers (within this GPU process)"""
+
+
+@define(frozen=True, kw_only=True)
+class QueryIDs:
+    """Wrapper for obs and var IDs from an |ExperimentAxisQuery|.
+
+    Serializable across multiple processes.
+    """
+
+    obs_joinids: NDArrayJoinId
+    """``obs`` row coordinates to read."""
+    var_joinids: NDArrayJoinId
+    """``var`` column coordinates to read."""
+    partition: Optional[Partition] = None
+    """GPU/Worker-partition info; typically populated by :meth:`partitioned`"""
+
+    @classmethod
+    def create(cls, query: ExperimentAxisQuery) -> "QueryIDs":
+        """Initialize a |QueryIDs| object from an |ExperimentAxisQuery|."""
+        obs_joinids = query.obs_joinids().to_numpy()
+        var_joinids = query.var_joinids().to_numpy()
+        return QueryIDs(
+            obs_joinids=obs_joinids,
+            var_joinids=var_joinids,
+        )
+
+    def split(
+        self,
+        *fracs: float,
+        seed: int | None = None,
+        method: SamplingMethod = "stochastic_rounding",
+    ) -> Tuple["QueryIDs", ...]:
+        """Split this |QueryIDs| into 1 or more |QueryIDs|, randomly sampled according ``fracs``.
+
+        - ``fracs`` must sum to $1$
+        - ``seed`` is optional
+        - ``method``: see |SamplingMethod| for details
+        """
+        split_fracs = np.cumsum(fracs)
+        assert fracs and np.isclose(split_fracs[-1], 1.0), "Fractions must sum to 1"
+
+        obs_joinids = self.obs_joinids
+        n_obs = len(obs_joinids)
+
+        rng = np.random.default_rng(seed)
+        shuffled_joinids = rng.permutation(obs_joinids)
+
+        if method == "deterministic":
```

Comment:
I believe I already mentioned this, but you will likely get requests for more types of splits (example: splits that are based upon the distribution of cells with a given metadata label, etc).  Would be good to think about how this might have to evolve to support that.

Example:  a 70/30 split by `cell_type`, meaning that for each unique cell_type value, 70% of the cells go into one split, 30% in the other.

Reply:
Oh, another random idea to toss into the pot:  just let the user pass a lamba that does their own split (i.e., returns the split indices).  As the user already has the experiment and query open, the can create a closure containing any required state.

Reply:
I don't think we need to resolve this issue in this PR - I just wanted to give you a heads up about a possible future request, and get some ideas down about how we might handle it.  Might be worth cloning this in a story?

Reply:
did this get filed as a story/issue, or punt?

Reply:
Filed as [sc-63967](https://app.shortcut.com/tiledb-inc/story/63967/tiledbsoma-ml-support-additional-split-types)